File: src/algorithms/strings/edit-distance/jaro-winkler-similarity/sources/jaro-winkler-similarity.py
# Jaro-Winkler Similarity
# Computes similarity between two strings using the Jaro formula,
# then boosts the score if the strings share a common prefix (up to 4 chars).
# Returns a value between 0.0 (completely dissimilar) and 1.0 (identical).
# Time: O(nm), Space: O(n)


def jaro_winkler_similarity(source: str, target: str) -> float:
    source_length = len(source)  # @step:initialize
    target_length = len(target)  # @step:initialize

    # Identical strings have similarity 1.0
    if source == target:  # @step:initialize
        return 1.0

    # Either empty string has similarity 0.0
    if source_length == 0 or target_length == 0:  # @step:initialize
        return 0.0

    # Match window: characters within this distance can be considered matching
    match_window = max(source_length, target_length) // 2 - 1  # @step:initialize

    source_matched = [False] * source_length  # @step:initialize
    target_matched = [False] * target_length  # @step:initialize
    match_count = 0  # @step:initialize

    # Find matching characters within the match window
    for source_idx in range(source_length):  # @step:compare
        window_start = max(0, source_idx - match_window)  # @step:compare
        window_end = min(target_length - 1, source_idx + match_window)  # @step:compare

        for target_idx in range(window_start, window_end + 1):  # @step:compare
            if not target_matched[target_idx] and source[source_idx] == target[target_idx]:  # @step:compare
                source_matched[source_idx] = True  # @step:compute-distance
                target_matched[target_idx] = True  # @step:compute-distance
                match_count += 1  # @step:compute-distance
                break

    # No matches means similarity is 0
    if match_count == 0:  # @step:compute-distance
        return 0.0

    # Count transpositions: matched chars in different order
    transposition_count = 0  # @step:compute-distance
    target_scan_idx = 0  # @step:compute-distance

    for source_idx in range(source_length):  # @step:compute-distance
        if not source_matched[source_idx]:  # @step:compute-distance
            continue

        while not target_matched[target_scan_idx]:  # @step:compute-distance
            target_scan_idx += 1  # @step:compute-distance

        if source[source_idx] != target[target_scan_idx]:  # @step:compute-distance
            transposition_count += 1  # @step:compute-distance

        target_scan_idx += 1  # @step:compute-distance

    # Jaro similarity formula
    half_transpositions = transposition_count / 2  # @step:compute-distance
    jaro_score = (  # @step:compute-distance
        match_count / source_length
        + match_count / target_length
        + (match_count - half_transpositions) / match_count
    ) / 3

    # Count common prefix length (up to 4 characters)
    max_prefix_length = 4  # @step:compute-distance
    prefix_length = 0  # @step:compute-distance

    for prefix_idx in range(min(max_prefix_length, source_length, target_length)):  # @step:compute-distance
        if source[prefix_idx] == target[prefix_idx]:  # @step:compute-distance
            prefix_length += 1  # @step:compute-distance
        else:
            break  # @step:compute-distance

    # Winkler bonus: reward common prefix
    winkler_bonus = prefix_length * 0.1 * (1 - jaro_score)  # @step:compute-distance
    jaro_winkler_score = jaro_score + winkler_bonus  # @step:compute-distance

    return round(jaro_winkler_score, 4)  # @step:complete

File: src/algorithms/strings/edit-distance/jaro-winkler-similarity/sources/JaroWinklerSimilarity.cpp
// Jaro-Winkler Similarity
// Computes similarity between two strings using the Jaro formula,
// then boosts the score if the strings share a common prefix (up to 4 chars).
// Returns a value between 0.0 (completely dissimilar) and 1.0 (identical).
// Time: O(nm), Space: O(n) where n and m are the string lengths.

#include <string>
#include <vector>
#include <algorithm>
#include <cmath>

double jaroWinklerSimilarity(const std::string& source, const std::string& target) {
    int sourceLength = static_cast<int>(source.length()); // @step:initialize
    int targetLength = static_cast<int>(target.length()); // @step:initialize

    // Identical strings have similarity 1.0
    if (source == target) return 1.0; // @step:initialize

    // Either empty string has similarity 0.0
    if (sourceLength == 0 || targetLength == 0) return 0.0; // @step:initialize

    // Match window: characters within this distance can be considered matching
    int matchWindow = std::max(sourceLength, targetLength) / 2 - 1; // @step:initialize

    std::vector<bool> sourceMatched(sourceLength, false); // @step:initialize
    std::vector<bool> targetMatched(targetLength, false); // @step:initialize

    int matchCount = 0; // @step:initialize

    // Find matching characters within the match window
    for (int sourceIdx = 0; sourceIdx < sourceLength; sourceIdx++) {
        // @step:compare
        int windowStart = std::max(0, sourceIdx - matchWindow); // @step:compare
        int windowEnd = std::min(targetLength - 1, sourceIdx + matchWindow); // @step:compare

        for (int targetIdx = windowStart; targetIdx <= windowEnd; targetIdx++) {
            // @step:compare
            if (!targetMatched[targetIdx] && source[sourceIdx] == target[targetIdx]) {
                // @step:compare
                sourceMatched[sourceIdx] = true; // @step:compute-distance
                targetMatched[targetIdx] = true; // @step:compute-distance
                matchCount++; // @step:compute-distance
                break;
            }
        }
    }

    // No matches means similarity is 0
    if (matchCount == 0) return 0.0; // @step:compute-distance

    // Count transpositions: matched chars in different order
    int transpositionCount = 0; // @step:compute-distance
    int targetScanIdx = 0; // @step:compute-distance

    for (int sourceIdx = 0; sourceIdx < sourceLength; sourceIdx++) {
        // @step:compute-distance
        if (!sourceMatched[sourceIdx]) continue; // @step:compute-distance

        while (!targetMatched[targetScanIdx]) {
            // @step:compute-distance
            targetScanIdx++; // @step:compute-distance
        }

        if (source[sourceIdx] != target[targetScanIdx]) {
            // @step:compute-distance
            transpositionCount++; // @step:compute-distance
        }

        targetScanIdx++; // @step:compute-distance
    }

    // Jaro similarity formula
    double halfTranspositions = transpositionCount / 2.0; // @step:compute-distance
    double jaroScore =
        (matchCount / static_cast<double>(sourceLength) // @step:compute-distance
            + matchCount / static_cast<double>(targetLength) // @step:compute-distance
            + (matchCount - halfTranspositions) / matchCount) // @step:compute-distance
        / 3.0; // @step:compute-distance

    // Count common prefix length (up to 4 characters)
    int maxPrefixLength = 4; // @step:compute-distance
    int prefixLength = 0; // @step:compute-distance

    for (int prefixIdx = 0; prefixIdx < std::min({maxPrefixLength, sourceLength, targetLength}); prefixIdx++) {
        // @step:compute-distance
        if (source[prefixIdx] == target[prefixIdx]) {
            // @step:compute-distance
            prefixLength++; // @step:compute-distance
        } else {
            break; // @step:compute-distance
        }
    }

    // Winkler bonus: reward common prefix
    double winklerBonus = prefixLength * 0.1 * (1.0 - jaroScore); // @step:compute-distance
    double jaroWinklerScore = jaroScore + winklerBonus; // @step:compute-distance

    return std::round(jaroWinklerScore * 10000.0) / 10000.0; // @step:complete
}
